### PR TITLE
Speed up SQLite.

### PR DESCRIPTION
**Pull request by AlCutter**
This speeds up SQLite operations considerably.

### CODE REVIEW

File: cpp/log/sqlite_db-inl.h [hunk 1]
```diff
@@ -17,8 +17,26 @@
...
+DEFINE_string(sqlite_temp_store, "DISK",
+              "Which SQLite temp_store option to use.");
+
+DEFINE_int32(sqlite_cache_size, 100000,
+             "Number of 1KB btree pages to keep in memory.");
+
+DEFINE_bool(sqlite_batch_into_transactions, true,
+            "Whether to batch operations into transactions behind the scenes. "
+            "Transactions will always be committed when new STHs are written. "
+            "It's recommended to run with the FULL synchronous mode when "
```
Comment (pphaneuf) [inline cpp/log/sqlite_db-inl.h]: I'm not sure this recommendation makes much sense? I think the synchronization mode only changes what happens when a transaction is completed, so it doesn't really matter what it does, it'll just be happening less often.

File: cpp/util/testing.cc [hunk 2]
```diff
@@ -5,6 +5,8 @@
 #include <glog/logging.h>
 #include <gtest/gtest.h>
 
+#include "log/logged_certificate.h"
```
Comment (pphaneuf) [inline cpp/util/testing.cc]: Huh?

File: cpp/log/sqlite_db-inl.h [hunk 3]
```diff
@@ -73,13 +91,50 @@ sqlite3* SQLiteOpen(const std::string& dbfile) {
...
+    if (FLAGS_sqlite_batch_into_transactions) {
+      LOG(WARNING) << "SQLite running with batched transactions, you should "
+                   << "set sqlite_synchronous_mode = FULL !";
+    }
   }
 
+  {
+    std::ostringstream oss;
+    oss << "PRAGMA journal_mode = " << FLAGS_sqlite_journal_mode;
+    sqlite::Statement statement(db_, oss.str().c_str());
```
Comment (pphaneuf) [inline cpp/log/sqlite_db-inl.h]: Ugh, ```sqlite::Statement``` should take a ```const string&``` for the statement, so you can build it easily, this is rather silly... For later, though.

File: cpp/log/sqlite_db-inl.h [hunk 4]
```diff
@@ -73,13 +91,50 @@ sqlite3* SQLiteOpen(const std::string& dbfile) {
...
+    CHECK_EQ(SQLITE_ROW, statement.Step());
+    std::string mode;
+    statement.GetBlob(0, &mode);
+    CHECK_STRCASEEQ(mode.c_str(), FLAGS_sqlite_journal_mode.c_str());
```
Comment (pphaneuf) [inline cpp/log/sqlite_db-inl.h]: You need ```.c_str()``` with this? I've never used it, just wondering...

File: cpp/log/sqlite_db-inl.h [hunk 5]
```diff
@@ -17,8 +17,26 @@
...
 // operation.
-DEFINE_bool(sqlite_synchronous_off, false,
-            "whether to set SQLite's \"synchronous\" to OFF");
+DEFINE_string(sqlite_synchronous_mode, "FULL",
+              "While SQLite synchronous option to use");
+
+DEFINE_string(sqlite_journal_mode, "WAL",
+              "Which SQLite journal_mode option to use.");
+
+DEFINE_string(sqlite_temp_store, "DISK",
```
Comment (pphaneuf) [inline cpp/log/sqlite_db-inl.h]: If that one is turning out to be pointless, might as well remove it?

**Comment by pphaneuf** [comment]
Done with this round of review.

File: cpp/log/sqlite_db-inl.h (hunk 1 above)
Comment (AlCutter) [inline cpp/log/sqlite_db-inl.h]: I suppose it's up to the operator how lucky they're feeling.
What I was trying to get at here was that the transaction batching amortises the cost of FULL synchronisation, and so there's less/no reason not to back away from it.
But maybe this isn't place; I'll remove this from here, and we should have some docs which discuss these flags.

File: cpp/util/testing.cc (hunk 2 above)
Comment (AlCutter) [inline cpp/util/testing.cc]: gnash gnash, that shouldn't be here. Removed.

File: cpp/log/sqlite_db-inl.h (hunk 4 above)
Comment (AlCutter) [inline cpp/log/sqlite_db-inl.h]: Yup, it's just a wrapper around `strcasecmp`

File: cpp/log/sqlite_db-inl.h (hunk 5 above)
Comment (AlCutter) [inline cpp/log/sqlite_db-inl.h]: Removed

Comment (AlCutter) [inline cpp/log/sqlite_db-inl.h]: Removed

File: cpp/util/testing.cc [hunk 6]
```diff
@@ -5,6 +5,7 @@
 #include <glog/logging.h>
 #include <gtest/gtest.h>
 
+
```
Comment (pphaneuf) [inline cpp/util/testing.cc]: Still a stray extra line?

File: cpp/server/ct-server.cc [hunk 7]
```diff
@@ -382,9 +382,10 @@ int main(int argc, char* argv[]) {
   CHECK(checker.LoadTrustedCertificates(FLAGS_trusted_cert_file))
       << "Could not load CA certs from " << FLAGS_trusted_cert_file;
 
-  if (FLAGS_sqlite_db == "")
+  if (FLAGS_sqlite_db == "") {
```
Comment (pphaneuf) [inline cpp/server/ct-server.cc]: While you're there, make this ```FLAGS_sqlite_db.empty()```? :wink:

File: cpp/server/ct-server.cc [hunk 8]
```diff
@@ -395,13 +396,14 @@ int main(int argc, char* argv[]) {
 
   Database<LoggedCertificate>* db;
 
-  if (FLAGS_sqlite_db != "")
+  if (FLAGS_sqlite_db != "") {
```
Comment (pphaneuf) [inline cpp/server/ct-server.cc]: Similar here, ```!FLAGS_sqlite_db.empty()```?

**Comment by pphaneuf** [comment]
LGTM, leaving those last things to your discretion, no need to review them separately.

File: cpp/log/sqlite_db-inl.h [hunk 9]
```diff
@@ -17,8 +17,21 @@
 // new node, it would do an initial load of its local database with
 // "synchronous" set to OFF, then put it back before starting normal
 // operation.
-DEFINE_bool(sqlite_synchronous_off, false,
-            "whether to set SQLite's \"synchronous\" to OFF");
+DEFINE_string(sqlite_synchronous_mode, "FULL",
+              "While SQLite synchronous option to use");
```
Comment (benlaurie) [inline cpp/log/sqlite_db-inl.h]: s/While/Which/?

Comment (benlaurie) [inline cpp/log/sqlite_db-inl.h]: Also, what are the options?

File: cpp/log/sqlite_db-inl.h [hunk 10]
```diff
@@ -17,8 +17,21 @@
...
+
+DEFINE_string(sqlite_journal_mode, "WAL",
+              "Which SQLite journal_mode option to use.");
```
Comment (benlaurie) [inline cpp/log/sqlite_db-inl.h]: Likewise, what are the options?

File: cpp/log/sqlite_db-inl.h (hunk 9 above)
Comment (AlCutter) [inline cpp/log/sqlite_db-inl.h]: Me fail English? That's unpossible!

Options are documented here: https://www.sqlite.org/pragma.html

I've added a comment and a hint in the flags' help strings

File: cpp/log/sqlite_db-inl.h (hunk 10 above)
Comment (AlCutter) [inline cpp/log/sqlite_db-inl.h]: same here.

**Pull request merged**